warehouse/slave.go

```diff
@@ -189,126 +207,170 @@ func (job *Payload) getDiscardsTable() string {
-					if err != nil {
-						uploadErrorChan <- err
-						return
-					}
-					loadFileUploadTimer.Since(loadFileUploadStart)
-					loadFileStats, err := os.Stat(uploadJob.outputFile.GetLoadFile().Name())
-					if err != nil {
-						uploadErrorChan <- err
-						return
-					}
-					loadFileOutputChan <- loadFileUploadOutput{
-						TableName:             tableName,
-						Location:              uploadOutput.Location,
-						ContentLength:         loadFileStats.Size(),
-						TotalRows:             jobRun.tableEventCountMap[tableName],
-						StagingFileID:         stagingFileId,
-						DestinationRevisionID: job.DestinationRevisionID,
-						UseRudderStorage:      job.UseRudderStorage,
+					jobStream <- &uploadJob{
+						tableName:  tableName,
+						outputFile: loadFile,
 					}
 				}
 			}
-		}(ctx)
+		}()
+		return jobStream
 	}
-	// Create upload jobs
-	go func() {
-		for tableName, loadFile := range jobRun.outputFileWritersMap {
-			uploadJobChan <- &loadFileUploadJob{tableName: tableName, outputFile: loadFile}
+
+	uploadLoadFile := func(uploader filemanager.FileManager, uploadFile encoding.LoadFileWriter, tableName string) (filemanager.UploadedFile, error) {
+		file, err := os.Open(uploadFile.GetLoadFile().Name())
+		if err != nil {
+			return filemanager.UploadedFile{}, fmt.Errorf("opening file: %w", err)
 		}
-	}()
+		defer file.Close()
+
+		var uploadLocation filemanager.UploadedFile
+		if slices.Contains(warehouseutils.TimeWindowDestinations, jr.job.DestinationType) {
+			uploadLocation, err = uploader.Upload(
+				ctx,
+				file,
+				warehouseutils.GetTablePathInObjectStorage(jr.job.DestinationNamespace, tableName),
+				jr.job.LoadFilePrefix,
+			)
+		} else {
+			uploadLocation, err = uploader.Upload(
+				ctx,
+				file,
+				jr.loadObjectFolder,
+				tableName,
+				jr.job.SourceID,
+				getBucketFolder(jr.job.UniqueLoadGenID, tableName),
+			)
+		}
+		return uploadLocation, err
+	}
+
+	process := func(uploadJobChan <-chan *uploadJob) <-chan *result {
+		processStream := make(chan *result, len(jr.outputFileWritersMap))
+
+		g, _ := errgroup.WithContext(ctx)
+		g.SetLimit(jr.numLoadFileUploadWorkers)
+
+		for i := 0; i < jr.numLoadFileUploadWorkers; i++ {
+			g.Go(func() error {
+				for uploadJob := range uploadJobChan {
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					default:
+						loadFileUploadStart := time.Now()
+
+						uploadOutput, err := uploadLoadFile(
+							uploader,
+							uploadJob.outputFile,
+							uploadJob.tableName,
+						)
+						if err != nil {
+							return fmt.Errorf("uploading load file: %w", err)
+						}
 
-	// Wait for response
-	for {
-		select {
-		case loadFileOutput := <-loadFileOutputChan:
-			loadFileUploadOutputs = append(loadFileUploadOutputs, loadFileOutput)
-			if len(loadFileUploadOutputs) == len(jobRun.outputFileWritersMap) {
-				return loadFileUploadOutputs, nil
+						mu.Lock()
```

Comment:
I found this code a bit confusing. First of all the `mu` name does not tell what variable it is supposed to protect. It is a common/good practice to do so. In this case it should be `totalUploadTimeMu`.

The other confusing thing is that we're protecting against writes only, perhaps leave a comment saying you expect to read the variable only once on defer after all routines are terminated.

Alternatively we could remove the double variable and just do `var totalUploadTime atomic.Duration`. Wdyt?

Reply:
Great point. Using `atomic.Duration`